### PR TITLE
File destination flush signal

### PR DESCRIPTION
**Pull request by shikharvashistha**
This pull introduces file rotation mechanism to syslog-ng

### CODE REVIEW

File: modules/file-rotation/file-rotation-parser.c [hunk 1]
```diff
@@ -33,7 +33,7 @@ static CfgLexerKeyword file_rotation_keywords[] =
 {
   { "file_rotation", KW_FILE_ROTATION },
   { "size", KW_SIZE },
-  { NULL }
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation-parser.c:36]: `NULL` marks the end of the list, it should not be removed.

File: modules/file-rotation/file-rotation.c [hunk 2]
```diff
@@ -27,6 +27,7 @@
 struct _FileRotationPlugin
 {
   LogDriverPlugin super;
+  char *interval;
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: Please use GLib types, wherever possible: `gchar *`

File: modules/file-rotation/file-rotation.c [hunk 3]
```diff
@@ -41,11 +42,21 @@ file_rotation_attach_to_driver(LogDriverPlugin *s, LogDriver *driver)
 static void
 file_rotation_free(LogDriverPlugin *s)
 {
-  FileRotationPlugin *self = (FileRotationPlugin *) s;
+  //FileRotationPlugin *self = (FileRotationPlugin *) s;
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: This does not belong in this commit.

Please make sure your commits are small atomic changes, where the commit message explains the whole change.

File: modules/file-rotation/file-rotation.h [hunk 4]
```diff
@@ -27,9 +27,11 @@
 #include "syslog-ng.h"
 
 typedef struct _FileRotationPlugin FileRotationPlugin;
+typedef struct _LogDriver LogDriver;
 
 FileRotationPlugin *file_rotation_new(void);
 
 void file_rotation_set_size(FileRotationPlugin *self, gsize size);
+void log_rotation_set_interval(LogDriver *self, char *interval);
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: We can use `FileRotationPlugin` as self. This plugin is not `LogDriver`-based.

File: modules/file-rotation/tests/test_file_rotation.c [hunk 5]
```diff
@@ -30,6 +30,7 @@ Test(file_rotation, test_example)
 {
   FileRotationPlugin *fr = file_rotation_new();
   file_rotation_set_size(fr, 100);
+  log_rotation_set_interval(fr, "1h");
```
Comment (MrAnno) [inline modules/file-rotation/tests/test_file_rotation.c]: I think we could use `logrotate`'s naming conventions here, for example:
daily, weekly, monthly, or yearly.

File: modules/affile/logproto-file-writer.c [hunk 6]
```diff
@@ -22,12 +22,7 @@
  */
 
 #include "logproto-file-writer.h"
-#include "signal-slot-connector/signal-slot-connector.h"
-#include "modules/http/http-worker.h"
-#include "modules/http/http-signals.h"
-#include "modules/http/http.h"
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: We allow (encourage) faking/squashing/rebasing commit history while working on a feature branch.
The goal here is to create a nice linear patchset for your pull request, that is easy to review.

For the same reason, please use `git rebase` instead of merge-based pulls when updating your working branch.

File: modules/affile/logproto-file-writer.c [hunk 7]
```diff
@@ -153,16 +148,16 @@ log_proto_file_writer_flush(LogProtoClient *s)
...
+    .filename = self->super.transport->name,
+    .size = self->buf_size, // implementation yet to be decided
+    .interval = 0, // implementation yet to be decided
   };
 
-  HTTPDestinationDriver *owner = (HTTPDestinationDriver *) self->worker->super.owner;
+  FileRotationDriver *owner = (FileRotationDriver *) self->owner;
 
-  EMIT(owner->super.super.super.super.signal_slot_connector, signal_http_header_request, &signal_data);
+  EMIT(owner->super.super.super.super.signal_slot_connector, signal_file_rotation_request, &signal_data);
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: This probably does not work due to the above reasoning.

File: modules/affile/logproto-file-writer.c [hunk 8]
```diff
@@ -36,7 +31,7 @@
 typedef struct _LogProtoFileWriter
 {
   LogProtoClient super;
-  HTTPDestinationWorker *worker;
+  LogThreadedDestWorker *owner;
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: We could store a `SignalSlotConnector *` pointer here, and it could be initialized by the constructor of LogProtoFileWriter: `log_proto_file_writer_new()`.

We need to pass the `AFFileDestDriver`'s `signal_slot_connector` instance to the constructor of LogProtoFileWriter.

File: modules/affile/logproto-file-writer.h [hunk 9]
```diff
@@ -25,6 +25,15 @@
...
 #include "logproto/logproto-client.h"
+#include "logthrdest/logthrdestdrv.h"
+#include "modules/file-rotation/file-rotation.h"
+
+typedef struct _FileRotationRequestData FileRotationRequestData;
+
+typedef struct
+{
+  LogThreadedDestDriver super;
+} FileRotationDriver;
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.h]: We don't need such a struct.

File: modules/affile/logproto-file-writer.h [hunk 10]
```diff
@@ -25,6 +25,15 @@
 #define LOG_PROTO_FILE_WRITER_H_INCLUDED
 
 #include "logproto/logproto-client.h"
+#include "logthrdest/logthrdestdrv.h"
+#include "modules/file-rotation/file-rotation.h"
+
+typedef struct _FileRotationRequestData FileRotationRequestData;
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.h]: Let's call this `FileFlushSignalData` for now.

The signal needs to be declared on this side, not in the file rotation plugin.
For that reason, let's call it `signal_file_flush`.

(We try to avoid mentioning file rotation here, we just provide generic signals, so any plugin can attach to them and do their job.)

All of these declarations should go into a separate header file with an LGPL license header.
We could call it `file-signals.h`.
An example can be found here: https://github.com/syslog-ng/syslog-ng/blob/master/modules/http/http-signals.h

**Comment by MrAnno** [comment]
Our commit message template is the following:

```
modulename: short description of the commit

optional longer description

Signed-off-by: Name <email>
```

Marking bug/fix/chore is a good practice, but we start with the module's name instead.
In our case, `affile` and `file-rotation`.

**Comment by shikharvashistha** [comment]
Hi, @MrAnno, I'm unable to see these conflicts locally don't know why these are there.

**Comment by MrAnno** [comment]
I can't review this. Please fix the git history.
I'm happy to help with git rebase or anything, we can talk about it on Gitter.

File: modules/file-rotation/file-rotation-parser.c [hunk 11]
```diff
@@ -33,6 +33,7 @@ static CfgLexerKeyword file_rotation_keywords[] =
 {
   { "file_rotation", KW_FILE_ROTATION },
   { "size", KW_SIZE },
+  { "interval", KW_INTERVAL },
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation-parser.c:36]: Adding the interval option does not belong to the flush-signal commit.

**Review by MrAnno — COMMENTED** [review]
I've also clicked "unresolve" on 3 of my previous notes. Those notes try to explain why this does not work.

Also, I can't compile the current version of this branch, `file-signals.h` is missing from this patchset, it is probably a rebase issue.

File: modules/affile/logproto-file-writer.c (hunk 7 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: So it should be http header request?

File: modules/affile/logproto-file-writer.c (hunk 8 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: I think I'm doing this [here](https://imgur.com/a/9WXptpE)

File: modules/file-rotation/file-rotation-parser.c (hunk 11 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation-parser.c:36]: I think these changes have been lost due to force push, what do you suggest should I separate out the commits ?

**Comment by shikharvashistha** [comment]
> I've also clicked "unresolve" on 3 of my previous notes. Those notes try to explain why this does not work.
> 
> Also, I can't compile the current version of this branch, `file-signals.h` is missing from this patchset, it is probably a rebase issue.

Yeah because it didn't got commited used git add -u

File: modules/affile/logproto-file-writer.c (hunk 8 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: > I think I'm doing this [here](https://imgur.com/a/9WXptpE)

[commit-line](https://github.com/MrAnno/syslog-ng/blob/5f73c4f8922ce92afaa01817204f9fd35d8c0192/modules/affile/logproto-file-writer.c#L36)

File: modules/affile/logproto-file-writer.c (hunk 7 above)
Comment (MrAnno) [reply modules/affile/logproto-file-writer.c]: No, it should be a `signal_file_rotation_request`, which needs to be declared in `file-signals.h`.

The problem here is that we need a `signal_slot_connector` instance, but not a completely new instance, we need `AFFileDestDriver`'s signal slot connector.
In order not to reference `AFFileDestDriver` directly from `LogProtoFileWriter` (unnecessary coupling), we could pass `AFFileDestDriver`'s signal slot connector to `log_proto_file_writer_new()` as a parameter.

File: modules/affile/logproto-file-writer.c (hunk 8 above)
Comment (MrAnno) [reply modules/affile/logproto-file-writer.c]: See my comment above. We need a specific instance from `AFFileDestDriver`, not a completely new one.

Also, please rename the field to `signal_slot_connector` in LogProtoFileWriter, as the connector is not the "owner" of the file writer.

File: modules/file-rotation/file-rotation-parser.c (hunk 11 above)
Comment (MrAnno) [reply modules/file-rotation/file-rotation-parser.c:36]: Yes, please create a separate commit.

As I mentioned:
We encourage faking commit history while working on a feature branch.
The goal here is to create a nice linear patchset for your pull request, that is easy to review.
The commit message should describe the change it is containing, which needs to be small/atomic, concentrating on one single thing.

File: modules/file-rotation/file-rotation.c [hunk 12]
```diff
@@ -22,29 +22,47 @@
...
-  LogDriverPlugin super;
-  gsize size;
-};
+
+  g_string_append_printf(self->filename, "%s", data->filename->str);
+}
 
 static gboolean
-file_rotation_attach_to_driver(LogDriverPlugin *s, LogDriver *driver)
+file_rotation_attach_to_driver(LogDriverPlugin *s, FileRotationPlugin *driver)
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: No need to change the signature of this function.

You can get a signal-slot-connector instance this way:

```
SignalSlotConnector *ssc = driver->super.signal_slot_connector;
```

(That instance will be the plugin's signal-slot-connector we attach to. In this example this is the `AFFileDestDriver`'s signal-slot-connector.)

File: modules/affile/logproto-file-writer.c (hunk 7 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: Fixed :+1:

File: modules/affile/logproto-file-writer.c (hunk 8 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: Updated :+1:

File: modules/affile/logproto-file-writer.h (hunk 10 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.h]: Created :+1:

File: modules/file-rotation/file-rotation.c (hunk 12 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: Changed :+1:

File: modules/file-rotation/file-rotation-parser.c (hunk 11 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation-parser.c:36]: Updated :+1:

File: modules/affile/file-opener.h [hunk 13]
```diff
@@ -64,7 +65,8 @@ struct _FileOpener
   LogTransport *(*construct_transport)(FileOpener *self, gint fd);
   LogProtoServer *(*construct_src_proto)(FileOpener *self, LogTransport *transport,
                                          LogProtoFileReaderOptions *proto_options);
-  LogProtoClient *(*construct_dst_proto)(FileOpener *self, LogTransport *transport, LogProtoClientOptions *proto_options);
+  LogProtoClient *(*construct_dst_proto)(FileOpener *self, LogTransport *transport, LogProtoClientOptions *proto_options,
+                                         SignalSlotConnector *connector);
```
Comment (MrAnno) [inline modules/affile/file-opener.h:69]: 👍🏻 

The `construct_dst_proto` function pointer is for imitating C++ virtual methods in C.
Since you changed the signature of the virtual method, you need to adjust the parameter list in subclasses too.

The compiler warnings show you the missing pieces, you will eventually reach `log_proto_file_writer_new()`.

`modules/affile/named-pipe.c` - we don't want to do anything with the connector here, just fix the signature
`modules/affile/regular-files.c` - this is where it gets interesting

File: modules/affile/file-opener.h [hunk 14]
```diff
@@ -28,6 +28,7 @@
 #include "compat/lfs.h"
 
 #include "file-perms.h"
+#include "lib/signal-slot-connector/signal-slot-connector.h"
```
Comment (MrAnno) [inline modules/affile/file-opener.h]: Let's just remove the `lib/` prefix, it should work too.

Comment (shikharvashistha) [reply modules/affile/file-opener.h]: Resolved

File: modules/affile/file-opener.h (hunk 13 above)
Comment (shikharvashistha) [reply modules/affile/file-opener.h:69]: Fixed :+1:

File: modules/affile/file-signals.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2022 Shikhar Vashistha
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
```
Comment (MrAnno) [inline modules/affile/file-signals.h]: We should change the license of this file to LGPL in my opinion.

For example:
https://github.com/syslog-ng/syslog-ng/blob/master/lib/cache.h#L4-L20

Please add this file to the copyright checker script too:
https://github.com/syslog-ng/syslog-ng/blob/master/tests/copyright/policy#L119

You can test it by running `make check-copyright`.

File: modules/affile/logproto-file-writer.c [hunk 16]
```diff
@@ -23,7 +23,8 @@
 
 #include "logproto-file-writer.h"
 #include "messages.h"
-
+#include "affile-dest.h"
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: That is not needed.

File: modules/affile/file-signals.h [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef FILE_SIGNALS_H_INCLUDED
+#define FILE_SIGNALS_H_INCLUDED
+
```
Comment (MrAnno) [inline modules/affile/file-signals.h:26]: #include "signal-slot-connector/signal-slot-connector.h"

File: modules/affile/logproto-file-writer.c [hunk 18]
```diff
@@ -32,6 +33,8 @@
 typedef struct _LogProtoFileWriter
 {
   LogProtoClient super;
+  SignalSlotConnector *signal_slot_connector;
+  AFFileDestDriver *signal_slot_connector_driver;
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: We should avoid storing or referencing `AFFileDestDriver` in this class, if possible.

File: modules/affile/logproto-file-writer.c [hunk 19]
```diff
@@ -134,6 +137,26 @@ log_proto_file_writer_flush(LogProtoClient *s)
...
+  };
+
+  //self->signal_slot_connector = self->signal_slot_connector_driver->super.super.super.signal_slot_connector;
+  //The flush-signal is not emitted on the AFFileDestDriver's signal-slot-connector, it is emitted using a completely different connector.
+  //The reason for this is that the AFFileDestDriver is not a LogProtoClient, but a LogProtoDestDriver.
+
+  //Register the signal to AFFileDestDriver's signal-slot-connector (the signal is emitted on the AFFileDestDriver's signal-slot-connector)
+
+  EMIT(self->signal_slot_connector, signal_file_flush, &signal_data);
+
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c:141]: Please remove all the comments here. We can get rid of the test data as well for now.
We can emit the signal with an empty `FileFlushSignalData`. The next task (and PR) will be about filling this data properly.

File: modules/affile/logproto-file-writer.c [hunk 20]
```diff
@@ -233,7 +255,11 @@ log_proto_file_writer_new(LogTransport *transport, const LogProtoClientOptions *
   LogProtoFileWriter *self = (LogProtoFileWriter *)g_malloc0(sizeof(LogProtoFileWriter) + sizeof(
       struct iovec)*flush_lines);
 
+
+
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c:242]: Those whitespaces seem unnecessary.

File: modules/affile/logproto-file-writer.h [hunk 21]
```diff
@@ -25,8 +25,9 @@
 #define LOG_PROTO_FILE_WRITER_H_INCLUDED
 
 #include "logproto/logproto-client.h"
+#include "modules/file-rotation/file-rotation.h"
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.h]: We should not include file-rotation in the affile module.

`#include "signal-slot-connector/signal-slot-connector.h"` should be enough here.

File: modules/file-rotation/file-rotation.c [hunk 22]
```diff
@@ -22,29 +22,46 @@
  */
 
 #include "file-rotation.h"
+#include "modules/http/http-signals.h"
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: We don't need HTTP signals here.

File: modules/file-rotation/file-rotation.c [hunk 23]
```diff
@@ -22,29 +22,46 @@
...
+#include "modules/affile/file-signals.h"
 #include "driver.h"
 
-struct _FileRotationPlugin
+static void
+_slot_append_file_rotation_request(FileRotationPlugin *self, FileFlushSignalData *data)
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: I'd call this function `_slot_file_rotation` or anything similar.

(We can't talk about "appending requests" here, those were just examples of the HTTP signaling).

File: modules/file-rotation/file-rotation.c [hunk 24]
```diff
@@ -22,29 +22,46 @@
...
 {
-  LogDriverPlugin super;
-  gsize size;
-};
+
+  g_string_append_printf(self->filename, "%s", data->filename->str);
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: I'd just add a `msg_trace()` here so you can test the slot by starting syslog-ng with `-Fdevt`.
There is no need to append the filename, we can just go with the msg_trace and nothing more for this PR.

File: modules/file-rotation/file-rotation.c [hunk 25]
```diff
@@ -22,29 +22,46 @@
...
+}
 
 static gboolean
 file_rotation_attach_to_driver(LogDriverPlugin *s, LogDriver *driver)
 {
-  msg_debug("File rotation plugin has been attached", evt_tag_str("to", driver->id));
+  SignalSlotConnector *ssc = driver->super.signal_slot_connector;
+
+  msg_debug("File rotation plugin has been attached", evt_tag_str("to", s->name));
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: The original `driver->id` was correct in my opinion, it showed the driver we attached our file-rotation plugin to.

File: modules/file-rotation/file-rotation.h [hunk 26]
```diff
@@ -25,11 +25,56 @@
 #define FILE_ROTATION_H_INCLUDED
 
 #include "syslog-ng.h"
+#include "driver.h"
+#include "signal-slot-connector/signal-slot-connector.h"
 
 typedef struct _FileRotationPlugin FileRotationPlugin;
+typedef struct _LogDriver LogDriver;
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: Please remove this.

File: modules/file-rotation/file-rotation.h [hunk 27]
```diff
@@ -25,11 +25,56 @@
...
+  FILE_ROTATION_SIZE,
+} FileRotationOption;
+
+typedef enum
+{
+  FILE_ROTATION_SUCCESS,
+  FILE_ROTATION_FAILURE,
+} FileRotationResult;
+
+struct _FileFlushSignalData
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: This should go into file-signals.h

File: modules/file-rotation/file-rotation.h [hunk 28]
```diff
@@ -25,11 +25,56 @@
...
+{
+  GString *filename;
+  gchar *interval;
+  gsize size;
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: filename can be useful, but we don't need interval and size here.

File: modules/file-rotation/file-rotation.h [hunk 29]
```diff
@@ -25,11 +25,56 @@
...
+  FileRotationResult result;
+  GString *filename;
+  gint64 last_rotation;
+  gint64 last_size_check;
+  gint64 last_interval_check;
+};
+
+
+
+#define signal_file_rotation_response SIGNAL(file_rotation, rotation_response, FileRotationResponseData *)
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: This is not needed.

File: modules/file-rotation/file-rotation.h [hunk 30]
```diff
@@ -25,11 +25,56 @@
...
+#include "signal-slot-connector/signal-slot-connector.h"
 
 typedef struct _FileRotationPlugin FileRotationPlugin;
+typedef struct _LogDriver LogDriver;
+typedef struct _FileRotationResponseData FileRotationResponseData;
+typedef enum
+{
+  FILE_ROTATION_INTERVAL,
+  FILE_ROTATION_SIZE,
+} FileRotationOption;
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.h]: This is unused, let's remove it.

File: modules/affile/file-signals.h (hunk 15 above)
Comment (shikharvashistha) [reply modules/affile/file-signals.h]: Updated :+1:

File: modules/affile/file-signals.h (hunk 17 above)
Comment (shikharvashistha) [reply modules/affile/file-signals.h:26]: :+1:

File: modules/affile/logproto-file-writer.c (hunk 18 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: :+1:

File: modules/affile/logproto-file-writer.c (hunk 19 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c:141]: :+1:

File: modules/affile/logproto-file-writer.c (hunk 20 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c:242]: :+1:

File: modules/affile/logproto-file-writer.h (hunk 21 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.h]: :+1:

File: modules/file-rotation/file-rotation.c (hunk 22 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: :+1:

File: modules/file-rotation/file-rotation.c (hunk 23 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: Updated :+1:

File: modules/file-rotation/file-rotation.c (hunk 24 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: ok sure

File: modules/file-rotation/file-rotation.c (hunk 25 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: updated

File: modules/file-rotation/file-rotation.h (hunk 26 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.h]: Removed

File: modules/file-rotation/file-rotation.h (hunk 27 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.h]: :+1:

File: modules/file-rotation/file-rotation.h (hunk 28 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.h]: Removed

File: modules/file-rotation/file-rotation.h (hunk 29 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.h]: :+1:

File: modules/file-rotation/file-rotation.h (hunk 30 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.h]: Removed :+1:

File: modules/affile/logproto-file-writer.c [hunk 31]
```diff
@@ -134,6 +135,14 @@ log_proto_file_writer_flush(LogProtoClient *s)
   self->buf_count = 0;
   self->sum_len = 0;
 
+  GString *request_body = g_string_new("");
+  g_string_append_printf(request_body, "%d", self->buf_size);
+  GString *file_name = g_string_new("");
+  g_string_append_printf(file_name, "%s", self->super.transport->name);
```
Comment (MrAnno) [inline modules/affile/logproto-file-writer.c]: `request_body` and `file_name` are not used currently. Let's remove them for now.

File: modules/file-rotation/file-rotation.c [hunk 32]
```diff
@@ -22,29 +22,44 @@
...
 #include "driver.h"
 
-struct _FileRotationPlugin
+static void
+_slot_file_rotation(FileRotationPlugin *self, FileFlushSignalData *data)
 {
-  LogDriverPlugin super;
-  gsize size;
-};
+  msg_trace("File rotation signal received", evt_tag_str("filename", self->filename->str));
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation.c]: Does not this crash?
Currently, self->filename is probably a NULL pointer.

**Review by MrAnno — COMMENTED** [review]
The proto constructor signature change hasn't been followed in our unit tests, please fix them there as well.

https://github.com/MrAnno/syslog-ng/runs/8036247829?check_suite_focus=true

File: modules/affile/logproto-file-writer.c (hunk 31 above)
Comment (shikharvashistha) [reply modules/affile/logproto-file-writer.c]: Updated 👍

File: modules/file-rotation/file-rotation.c (hunk 32 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation.c]: Updated 👍

File: modules/file-rotation/file-rotation-grammar.ym [hunk 33]
```diff
@@ -80,6 +81,7 @@ file_rotation_options
 
 file_rotation_option
   : KW_SIZE '(' nonnegative_integer ')' { file_rotation_set_size(last_file_rotation_plugin, $3); }
+  | KW_INTERVAL '(' string ')' { log_rotation_set_interval(last_file_rotation_plugin, $3); free($3); }
```
Comment (MrAnno) [inline modules/file-rotation/file-rotation-grammar.ym]: The PR still can not be built. `log_rotation_set_interval` -> `file_rotation_set_interval`

**Comment by MrAnno** [comment]
There is one remaining CMake build issue in the CI:

```
[ 80%] Building C object syslog-ng-ctl/CMakeFiles/syslog-ng-ctl.dir/syslog-ng-ctl.c.o
/__w/syslog-ng/syslog-ng/modules/file-rotation/file-rotation.c:25:10: fatal error: 'modules/affile/file-signals.h' file not found
#include "modules/affile/file-signals.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

You can fix this by adding the following to `modules/file-rotation/CMakeLists.txt`:

```diff
--- a/modules/file-rotation/CMakeLists.txt
+++ b/modules/file-rotation/CMakeLists.txt
@@ -15,6 +15,7 @@ add_module(
   TARGET file_rotation
   GRAMMAR file-rotation-grammar
   SOURCES ${FILE_ROTATION_SOURCES}
+  INCLUDES ${PROJECT_SOURCE_DIR}
 )
```

File: modules/file-rotation/file-rotation-grammar.ym (hunk 33 above)
Comment (shikharvashistha) [reply modules/file-rotation/file-rotation-grammar.ym]: The file where I made changes couldn't be pushed due to git issues.

**Comment by shikharvashistha** [comment]
> There is one remaining CMake build issue in the CI:
> 
> ```
> [ 80%] Building C object syslog-ng-ctl/CMakeFiles/syslog-ng-ctl.dir/syslog-ng-ctl.c.o
> /__w/syslog-ng/syslog-ng/modules/file-rotation/file-rotation.c:25:10: fatal error: 'modules/affile/file-signals.h' file not found
> #include "modules/affile/file-signals.h"
>          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> You can fix this by adding the following to `modules/file-rotation/CMakeLists.txt`:
> 
> ```diff
> --- a/modules/file-rotation/CMakeLists.txt
> +++ b/modules/file-rotation/CMakeLists.txt
> @@ -15,6 +15,7 @@ add_module(
>    TARGET file_rotation
>    GRAMMAR file-rotation-grammar
>    SOURCES ${FILE_ROTATION_SOURCES}
> +  INCLUDES ${PROJECT_SOURCE_DIR}
>  )
> ```

Yeah I've added it to CMakeLists.txt

**Review by MrAnno — APPROVED** [review]
